libcxx/test/std/algorithms/alg.sorting/alg.heap.operations/make.heap/ranges_make_heap.pass.cpp

```diff
@@ -60,6 +60,13 @@ static_assert(!HasMakeHeapR<UncheckedRange<const int*>>); // Doesn't satisfy `so
 
 template <std::size_t N, class T, class Iter>
 constexpr void verify_heap(const std::array<T, N>& heapified, Iter last, std::array<T, N> expected) {
+  assert(heapified == expected);
```

Comment:
This looks wrong to me. We should keep the iterator and range overload tests equivalent. Same for a few cases below.

Reply:
I made some efforts trying to unify those two sets of tests, but since `contiguous_iterator` is only designed to hold pointers (as is suggested here: [test_iterators.h#L344](https://github.com/llvm/llvm-project/blob/main/libcxx/test/support/test_iterators.h#L344)), I couldn't find an easy way to do that. I ended up deleting that static assertion and made some changes to `contiguous_iterator`.
Those changes did not break any tests, but since we use this class quite extensively, I would like your opinion on whether those changes make sense to the rest of the test suite.

Reply:
I don't think we should remove that static assertion. I'm also not sure what the problem was you were trying to solve with that change. I you need a range you can just do `std::ranges::subrange(Iter(array.data()), Sent(array.data() + array.size()))`.

Reply:
For `last` in `verify_heap`, we get `contiguous_iterator<int*>` in some cases and `std::array::iterator` in some other cases, and we later on want to compare it with `heapified.end()`. Since there isn't a strictly conforming way to cast `std::array::iterator` to pointers (correct me if I'm wrong), the two solutions I was to either allow `contiguous_iterator` to contain `std::array::iterator`, or to check equality in different ways, as illustrated in [e69008d](https://github.com/llvm/llvm-project/pull/70729/commits/e69008dcd6965ecb723294136a26c343e7092b39#diff-3d1c9b09e86d6dace3fe63664d13c76632c4127dd690b9124f61c28af8130d03L61-L66).

In hindsight, the naming in [e69008d](https://github.com/llvm/llvm-project/pull/70729/commits/e69008dcd6965ecb723294136a26c343e7092b39#diff-3d1c9b09e86d6dace3fe63664d13c76632c4127dd690b9124f61c28af8130d03L61-L66) could be misleading. If the general idea of it looks good to you, I can refine it as below:
```cpp
template <std::size_t N, class T>
constexpr void verify_last(const std::array<T, N>& heapified, T *last) {
  assert(last == heapified.data() + heapified.size());
}

template <std::size_t N, class T>
constexpr void verify_last(const std::array<T, N>& heapified, const typename std::array<T, N>::iterator& last) {
  assert(last == heapified.end());
}

template <std::size_t N, class T, class Iter>
constexpr void verify_heap(const std::array<T, N>& heapified, Iter last, const std::array<T, N>& expected) {
  assert(heapified == expected);
  verify_last(heapified, base(last));
  assert(std::is_heap(heapified.begin(), heapified.end()));
}
```

Reply:
You can use `std::to_address`, but I think we should instead simply remove the uses of `array::iterator`, since we care about the algorithm and not the container.

Reply:
I agree that we should use pointers as much as possible for these tests. In practice, this might require a bigger rewriting of the tests. For now, I have updated `verify_heap` so that we are _eventually_ checking equality of pointers.

Reply:
@philnik777 Do the new changes look good to you?